# Algorithmes gloutons
## 1 . rendu de monnaie

In [1]:
liste_monnaie=[200,100,50,20,10,5,2,1]
def rendu_monnaie(somme,lp):
    liste_pieces_a_rendre = []
    nb_piece=0
    for piece in lp:                 #pour chacune des pièces de la liste
        nb_piece = somme // piece    #quotient de la division : donne le nombre de pièce de ce type qu'on peut rendre
        somme = somme % piece        #reste de la division : la monnaie pas rendu après avoir traitée cette pièce
        for i in range(nb_piece):
            liste_pieces_a_rendre.append(piece) 
    return(liste_pieces_a_rendre)

In [2]:
print(rendu_monnaie(150,liste_monnaie))

[100, 50]


In [3]:
print(rendu_monnaie(388,liste_monnaie))

[200, 100, 50, 20, 10, 5, 2, 1]


## 2. problème du sac à dos
### a. structure de donnée

In [4]:
# structure de données
# chaque objet sera un tableau [nom,valeur,masse]
# la liste des objets sera une liste d'objet
liste_objets=[
        ["lingot",10,1],
        ["Imac",3,6],
        ["smartTV",4,8],
        ["masque",3,1]
]
print(liste_objets)

[['lingot', 10, 1], ['Imac', 3, 6], ['smartTV', 4, 8], ['masque', 3, 1]]


### b.première stratégie

<div style="font-size:1.3em">
Dans l'algorithme, il faut trier les objets par valeurs. Ce n'était pas demandé dans l'exercice, on pouvait le faire manuellement. <br>Mais on peut le faire facilement avec la fonction tri_sur_valeur ci-dessous.
<br><br>   <u>Explication</u><br><br>
liste.sort : tri la liste en place, c'est-à-dire que la liste est modifiée.
<ul>    
<li>key indique sur quelle clé trier </li>
<li>lambda est une fonction anonyme (sans nom) <br>
     celle définie ici est équivalente à :<br>
    <em>
    def sansnom(liste):<br>
        &nbsp;&nbsp;&nbsp;&nbsp;  return liste[1]<br></em>
    on remarque qu'avec le mot clé lambda, la syntaxe est simplifiée : pas de mot clé def et return et pas de nom.<br>
    Elle renvoie donc pour chaque élément de la liste le second item<br>
   <em> (rappel : les indices commencent à zéro donc 1 est le deuxième)</em></li>
<li>reverse=True : indique qu'il faut trier dans l'ordre décroissant   </li> 
</ul>
</div>

In [5]:
def tri_valeur(liste):
    return liste.sort(key=lambda liste: liste[1], reverse=True )

tri_valeur(liste_objets)
print(liste_objets)

[['lingot', 10, 1], ['smartTV', 4, 8], ['Imac', 3, 6], ['masque', 3, 1]]


In [6]:
def sac_a_dos(liste_objets, masse_disponible):
    tri_valeur(liste_objets)                      # on trie les objets par valeur
    objets_sac = []                               # tableau contenant les objets du sac
    masse_sac = 0                                
    valeur_sac = 0
    for objet in liste_objets:                    # pour chacun des objets disponibles
        if objet[2] <= masse_disponible:          # si la masse de l'objet est inférieure à la masse encvore disponible
            objets_sac.append(objet)              # on ajoute l'objet dans le sac
            masse_disponible -= objet[2]          # la masse disponible est dimunée de la masse de l'objet ajouté
            masse_sac += objet[2]                 # la masse du sac est augmentée de la masse de l'objet
            valeur_sac+=objet[1]                  # la valeur du sac est augmentée de la valeur de l'objet 
    return objets_sac,valeur_sac,masse_sac

In [7]:
print (sac_a_dos(liste_objets,8))

([['lingot', 10, 1], ['Imac', 3, 6], ['masque', 3, 1]], 16, 8)


In [8]:
print (sac_a_dos(liste_objets,15))

([['lingot', 10, 1], ['smartTV', 4, 8], ['Imac', 3, 6]], 17, 15)


In [9]:
print (sac_a_dos(liste_objets,4))

([['lingot', 10, 1], ['masque', 3, 1]], 13, 2)


### c.seconde stratégie
<div style="font-size:1.3em">
   Dans l'algorithme, il faut trier les objets selon les rapports valeur/masse.
La fonction tri_valeur_sur_masse ci-dessous le fait facilement.
Il suffit de modifier la fonction lambda.
   </div>

In [10]:

def tri_valeur_sur_masse(liste):
    return liste.sort(key=lambda liste: liste[1]/liste[2], reverse=True )

tri_valeur_sur_masse(liste_objets)
print(liste_objets)

[['lingot', 10, 1], ['masque', 3, 1], ['smartTV', 4, 8], ['Imac', 3, 6]]


In [11]:
print (sac_a_dos(liste_objets,8))

([['lingot', 10, 1], ['masque', 3, 1], ['Imac', 3, 6]], 16, 8)


In [12]:
print (sac_a_dos(liste_objets,15))

([['lingot', 10, 1], ['smartTV', 4, 8], ['masque', 3, 1]], 17, 10)


In [13]:
print (sac_a_dos(liste_objets,4))

([['lingot', 10, 1], ['masque', 3, 1]], 13, 2)


### d. comparaison des stratégie
<div style="font-size:1.3em">
La <b>première stratégie</b> donne pour un sac pouvant accepter 15 kg: <br>
lingot, smartTV, Imac soit une valeur de 17000€ pour  15kg<br>
alors que la <b>seconde</b> donne :<br>
lingot, smartTV,masque soit une valeur de 17000€ pour 10 kg<br>
<br>
<b>Dans ce cas, la seconde stratégie est la meilleure.</b>
</div>

## 3. Sac à dos avec un échantillon plus grand

In [14]:
liste_objets=[
    ["A",6,20],["B",8,30],["C",14,50],["D",6,20],["E",13,40],["F",17,60],["G",10,30],
    ["H",4,10],["I",5,14],["J",11,36],["K",26,72],["L",35,86],["M",2,5],["N",1,3],
    ["O",2,7],["P",7,23],["Q",15,49],["R",17,57], ["S",30,69], ["T",3,12]
]
print(liste_objets)

[['A', 6, 20], ['B', 8, 30], ['C', 14, 50], ['D', 6, 20], ['E', 13, 40], ['F', 17, 60], ['G', 10, 30], ['H', 4, 10], ['I', 5, 14], ['J', 11, 36], ['K', 26, 72], ['L', 35, 86], ['M', 2, 5], ['N', 1, 3], ['O', 2, 7], ['P', 7, 23], ['Q', 15, 49], ['R', 17, 57], ['S', 30, 69], ['T', 3, 12]]


### a.première stratégie

In [15]:
tri_valeur(liste_objets)
print(liste_objets)

[['L', 35, 86], ['S', 30, 69], ['K', 26, 72], ['F', 17, 60], ['R', 17, 57], ['Q', 15, 49], ['C', 14, 50], ['E', 13, 40], ['J', 11, 36], ['G', 10, 30], ['B', 8, 30], ['P', 7, 23], ['A', 6, 20], ['D', 6, 20], ['I', 5, 14], ['H', 4, 10], ['T', 3, 12], ['M', 2, 5], ['O', 2, 7], ['N', 1, 3]]


In [16]:
print (sac_a_dos(liste_objets,15))

([['I', 5, 14]], 5, 14)


In [17]:
print (sac_a_dos(liste_objets,21))

([['A', 6, 20]], 6, 20)


In [18]:
print (sac_a_dos(liste_objets,48))

([['E', 13, 40], ['M', 2, 5], ['N', 1, 3]], 16, 48)


In [19]:
print (sac_a_dos(liste_objets,90))

([['L', 35, 86], ['N', 1, 3]], 36, 89)


In [20]:
print (sac_a_dos(liste_objets,692))

([['L', 35, 86], ['S', 30, 69], ['K', 26, 72], ['F', 17, 60], ['R', 17, 57], ['Q', 15, 49], ['C', 14, 50], ['E', 13, 40], ['J', 11, 36], ['G', 10, 30], ['B', 8, 30], ['P', 7, 23], ['A', 6, 20], ['D', 6, 20], ['I', 5, 14], ['H', 4, 10], ['T', 3, 12], ['M', 2, 5], ['O', 2, 7]], 231, 690)


### b.deuxième statégie

In [21]:
tri_valeur_sur_masse(liste_objets)
print(liste_objets)

[['S', 30, 69], ['L', 35, 86], ['H', 4, 10], ['M', 2, 5], ['K', 26, 72], ['I', 5, 14], ['G', 10, 30], ['N', 1, 3], ['E', 13, 40], ['Q', 15, 49], ['J', 11, 36], ['P', 7, 23], ['A', 6, 20], ['D', 6, 20], ['R', 17, 57], ['O', 2, 7], ['F', 17, 60], ['C', 14, 50], ['B', 8, 30], ['T', 3, 12]]


In [22]:
print (sac_a_dos(liste_objets,15))

([['I', 5, 14]], 5, 14)


In [23]:
print (sac_a_dos(liste_objets,21))

([['A', 6, 20]], 6, 20)


In [24]:
print (sac_a_dos(liste_objets,48))

([['E', 13, 40], ['M', 2, 5], ['N', 1, 3]], 16, 48)


In [25]:
print (sac_a_dos(liste_objets,90))

([['L', 35, 86], ['N', 1, 3]], 36, 89)


In [26]:
print (sac_a_dos(liste_objets,692))

([['L', 35, 86], ['S', 30, 69], ['K', 26, 72], ['R', 17, 57], ['F', 17, 60], ['Q', 15, 49], ['C', 14, 50], ['E', 13, 40], ['J', 11, 36], ['G', 10, 30], ['B', 8, 30], ['P', 7, 23], ['A', 6, 20], ['D', 6, 20], ['I', 5, 14], ['H', 4, 10], ['T', 3, 12], ['M', 2, 5], ['O', 2, 7]], 231, 690)


### c.comparaison des deux méthodes
<div style="font-size:1.3em">
<table>
     <tr><th>Masse du sac</th><th colspan=2>Première stratégie</th><th colspan=2>Deuxième stratégie</th></tr>
    <tr><th></th><th>Valeur</th><th>Masse</th><th>Valeur</th><th>Masse</th></tr>
     <tr><td>15</td><td style="color:red">5</td><td style="color:red">14</td><td style="color:green">6</td><td style="color:green">15</td></tr>
     <tr><td>21</td><td style="color:red">6</td><td style="color:red">20</td><td style="color:green">7</td><td style="color:green">18</td></tr>
     <tr><td>48</td><td style="color:green">16</td><td style="color:green">48</td><td style="color:red">14</td><td style="color:red">39</td></tr>
     <tr><td>90</td><td style="color:red">36</td><td style="color:red">89</td><td style="color:green">37</td><td style="color:green">87</td></tr>
     <tr><td>692</td><td style="color:green">231</td><td  style="color:green">690</td><td style="color:red">229</td><td style="color:red">681</td></tr>
</table>    
En vert, la meilleure des solutions et en rouge la moins bonne.<br>
 <br>   <u> Conclusion</u> :<br> Aucune des deux méthodes ne donnent dans tous les cas la solution idéale.<br>
Nous verrons donc en Terminale une autre approche de ce problème pour obtenir dans tous les cas la meilleure solution.
 </div>